container-search/src/main/java/com/yahoo/search/dispatch/rpc/RpcProtobufFillInvoker.java

```diff
@@ -135,34 +143,73 @@ private void sendDocsumsRequest(int nodeId, List<FastHit> hits, byte[] payload,
             log.warning("Got hits with node id " + nodeId + ", which is not included in the current dispatch config");
             return;
         }
-
         Query query = result.getQuery();
         Compressor.Compression compressionResult = compressor.compress(query, payload);
         node.request(RPC_METHOD, compressionResult.type(), payload.length, compressionResult.data(),
                 roe -> receive(roe, hits), clientTimeout);
     }
 
+    private ResponseAndHits getNextResponse(Query query) throws InterruptedException {
+        long timeLeftMs = query.getTimeLeft();
+        if (timeLeftMs <= 0) {
+            return null;
+        }
+        var responseAndHits = responses.poll(timeLeftMs, TimeUnit.MILLISECONDS);
+        if (responseAndHits == null || responseAndHits.response().timeout()) {
+            return null;
+        }
+        return responseAndHits;
+    }
+
     private void processResponses(Result result, String summaryClass) throws TimeoutException {
         try {
-            int skippedHits = 0;
+            List<FastHit> skippedHits = new ArrayList<>();
             while (outstandingResponses > 0) {
-                long timeLeftMs = result.getQuery().getTimeLeft();
-                if (timeLeftMs <= 0) {
-                    throwTimeout();
-                }
-                var responseAndHits = responses.poll(timeLeftMs, TimeUnit.MILLISECONDS);
+                var responseAndHits = getNextResponse(result.getQuery());
                 if (responseAndHits == null) {
                     throwTimeout();
                 }
-                var response = responseAndHits.getFirst();
-                if (response.timeout()) {
-                    throwTimeout();
-                }
-                var hitsContext = responseAndHits.getSecond();
-                skippedHits += processResponse(result, response, hitsContext, summaryClass);
+                skippedHits.addAll(processOneResponse(result, responseAndHits, summaryClass, false));
                 outstandingResponses--;
             }
-            if (skippedHits != 0) {
+            if (skippedHits.isEmpty()) {
+                // all done OK
+                return;
+            }
+            int numSkipped = skippedHits.size();
```

Comment:
I suggest documenting why retrying docsum fetch is needed in some cases to better explain the following code. Also consider putting all of the retrying in a separate function to make it easier to read processResponses().